agent/resources/resources_linux.go

```diff
@@ -0,0 +1,120 @@
+// +build linux
+
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package resources
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/resources/cgroup"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+// cgroupWrapper implements the Resource interface
+type cgroupWrapper struct {
+	control cgroup.Control
+}
+
+// New is used to return an object that implements the Resource interface
+func New() Resource {
+	return newResources(cgroup.New())
+}
+
+func newResources(control cgroup.Control) Resource {
+	return &cgroupWrapper{
+		control: control,
+	}
+}
+
+// Init is used to initialize the resource
+func (c *cgroupWrapper) Init() error {
+	return c.cgroupInit()
+}
+
+// Setup sets up the resource
+func (c *cgroupWrapper) Setup(task *api.Task) error {
+	return c.setupCgroup(task)
+}
+
+// Cleanup removes the resource
+func (c *cgroupWrapper) Cleanup(task *api.Task) error {
+	return c.cleanupCgroup(task)
+}
+
+// cgroupInit is used to create the root '/ecs/ cgroup
+func (c *cgroupWrapper) cgroupInit() error {
+	if c.control.Exists(config.DefaultTaskCgroupPrefix) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", config.DefaultTaskCgroupPrefix)
+		return nil
+	}
+	return c.control.Init()
+}
+
+// setupCgroup is used to create the task cgroup
+func (c *cgroupWrapper) setupCgroup(task *api.Task) error {
+	cgroupRoot, err := task.BuildCgroupRoot()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to determine cgroup root for task: %s", task.Arn)
+	}
+
+	seelog.Debugf("Setting up cgroup at: %s for task: %s", cgroupRoot, task.Arn)
+
+	if c.control.Exists(cgroupRoot) {
+		seelog.Debugf("Cgroup at %s already exists, skipping creation", cgroupRoot)
+		return nil
+	}
+
+	linuxResourceSpec, err := task.BuildLinuxResourceSpec()
+	if err != nil {
+		return errors.Wrapf(err, "resource: setup cgroup: unable to build resource spec for task: %s", task.Arn)
+	}
+
+	cgroupSpec := cgroup.Spec{
+		Root:  cgroupRoot,
+		Specs: &linuxResourceSpec,
+	}
+
+	cgrp, err := c.control.Create(&cgroupSpec)
+	if err != nil {
```

Comment:
I wonder if we need to think about the use-case where there could be an error in one of the subsystems after a number of subsystems have already been initialized (https://github.com/containerd/cgroups/blob/master/cgroup.go#L73). How do we clean up in such scenarios?

Reply:
The `cgroup` subsystem and associated controllers have been around for a long time and are very efficient. I don't see any particular reason why a specific `subsystem/controller` would misbehave. I don't think we have to concern ourselves with this problem as we don't operate at such a low level within the system. 

If you have any specific concerns, I'd like to hear more about them.

Reply:
I am not advocating that we change anything here, but I don't buy the argument that something has been around for a long time and can thus be considered to never error out. We're interacting with the cgroup system using a go library and any number of things can go wrong in that interaction. It's a very useful exercise to look at what the library is actually doing and evaluate if something could go wrong and how we respond to scenarios when it does go wrong. 

Have we verified that the call paths are all thread safe and do not have any blocking routines before making that statement? Are we sure that there are no additional safeguards that we should be putting in place? We have seen such things coming back to bite us when using other 3p packages. 

As an example, what happens if a customer has a configuration where they've disabled/unmounted a particular cgroup system? Or, how does this library behave if the Agent is launched with reduced privileges? Again, these things might not necessarily result in the error in this particular method, but I'm not sure if all of these scenarios will be/are being gracefully handled.

Reply:
Oh, the scenario that I had in mind when I asked you this was let's say Agent gets restarted after a couple of subsystems have been initialized. What happens on recovery?

Reply:
I implied that the cgroup subsystem within the Linux kernel is fairly stable and hasn't changed much in a long time. There are plans for `cgroups v2` but its still `WIP` and is missing  controllers last time I investigated. 

I have looked into the 3p library we are using and its well maintained and doesn't have any of the problems you are concerned about. I have experimented with the library and have not identified any major issues. Even when specific cgroup subsystems are disabled, the library still works accurately. 

The features we care about depend primarily on the `cpu` and `memory` subsystems. 

`amazon-ecs-init` is going to launch `amazon-ecs-agent` with `CAP_NET_ADMIN` and `CAP_SYS_ADMIN`. 

If the instance has a working docker environment and the agent container is launched with the required mounts, the library and the agent work in harmony to launch tasks accurately.

Reply:
If the agent is restarted and new cgroup subsystems have been initialized, the agent should still function without any issues. The agent would attempt to `Load` the cgroup if it already exists before making any attempts to `recreate` the cgroup.

Reply:
Cgroup subsystems can be enabled and disabled independently.  However, Docker generally relies on the same ones being enabled that we rely on (`cpu` and `memory` here), so I'm not too worried about that case.